### DISCUSSION TITLE
Leveraging in an OEM managed package?

### DISCUSSION THREAD

**Discussion by KeithClarke** [jamessimone/apex-rollup#359] [Q&A] [answered]
Hi,

This looks great to me, and I wanted to give it a go in a scratch org, but that was blocked by this:

> You cannot install an unlocked package without a namespace into an org with a namespace

because the scratch org had a namespace defined. I can use a different org to try it out so no biggie there.

But this does prompt the question of how to leverage apex-rollup in a managed package where the managed package development is done in a namespaced scratch org? A namespaced managed package of apex-rollup would likely solve that, but that is not something you have chosen to provide?

**Comment by jamessimone** [comment]
@KeithClarke yes, I agree that a namespaced managed package version would probably be the best option here. Let me do some investigation on this over the next week. In the meantime, let me know if you have other questions, and thanks for reaching out!

**Reply by KeithClarke** [reply]
Thanks James. Maybe e.g. Salesforce security reviews would be an unwelcome burden. And being able to work straight away without a user haven't to configure is important, so e.g. providing an override to the custom setting in the CMDT would allow that - relates to #360.

**Reply by jamessimone** [reply]
@KeithClarke I was out of town this weekend and wasn't thinking clearly when answering this initially - it's possible to provide namespaced unlocked packages (2GP), which takes quite a bit out of the bite of providing this functionality. I'll be prototyping the namespaced version of Apex Rollup this week and hope to release that over the next few days.

**Reply by jamessimone** [reply]
@KeithClarke just [finished the work necessary to add a namespaced version of the package](https://github.com/jamessimone/apex-rollup/commit/6cea2e2d15422f9ee5ca7020b0b2f05e8bf094a5). I'll be looking to release this in the next day or two. Thank you!

**Reply by jamessimone** [reply]
@KeithClarke #362  put the first beta package for the namespaced version of Apex Rollup out there. I’ll be doing additional work next week on the next version to get it ready for production for real. I’ll keep you updated!

**Reply by jamessimone** [reply]
@KeithClarke the namespaced version of the package is now available! Please see the [namespaced package's README](https://github.com/jamessimone/apex-rollup/tree/main/rollup-namespaced) for further info.

**Comment by jamessimone** [comment, marked as answer]
The [namespaced version of the package](https://github.com/jamessimone/apex-rollup/tree/main/rollup-namespaced) is now available!

**Reply by pyao-bwc** [reply]
We're using the non-namespaced package right now. 

When we upgrade apex-rollup, Custom Metadata Type records that are included with the package are overwritten (as Salesforce warns on the CMT record page before editing).

I keep forgetting to re-enable Logging on the ``Org Default`` control record, and we deploy these CMT changes through our pipeline, so it's not just a matter of checking the box.

I thought I could make an ``Enabled`` control record that's custom to us that wouldn't get overwritten (which would be nice if we start changing the other values on the record). However, when creating the Plugin CMT record to enable the ``Rollup Logger`` plugin, I realized the Developer Name has to be unique, and it also has to be the name of the class, so we can't have two ``RollupLogger`` plugins and ``apex-rollup`` will ship with one of them.

If we use the namespaced version of the package we could create our own RollupLogger plugin to use for our own control record, so that sounds good.

Are there considerations to using the namespaced version (aside from the ones in [the readme](https://github.com/jamessimone/apex-rollup/tree/main/rollup-namespaced))?

If we had multiple control records (maybe we'd use a control record to group rollups that are also scheduled, or rollups that had different batch sizes, etc), we'd be stuck, but we might not need that.

**Reply by pyao-bwc** [reply]
Hi @jamessimone , I had trouble using the namespaced package and our custom logger that extends RollupLogger. 

If I try to create a class like this:
```
public with sharing class ApexRollupLogger extends please.RollupLogger {
```
I get:
```
Type is not visible: please.RollupLogger
```

Is that because RollupLogger is defined as public and not global?

We could implement ILogger instead of extending RollupLogger (though we like ``RollupLogger.getLogStringFromObject`` and would have to copy paste it). But ILogger is public instead of global as well, if that's the problem.

**Reply by jamessimone** [reply]
@pyao-bwc I can raise the visibility of that class in the next version. You've brought up some challenges with keeping logging enabled that I'll have to think over - it almost seems like there's two separate needs here that might help:

- ability to stipulate a Rollup Plugin Parameter which suppresses a given plugin - this gives you the ability to fine-tune which loggers end up being initialized in a way that's compatible with CI
- ability to stipulate "always on" loggers via a Plugin Parameter which allows you to bypass which Rollup Control a given plugin is associated with

Let me know what you think!

**Reply by pyao-bwc** [reply]
Thanks @jamessimone ! 

What if we put the class name in another field on the RollupPlugin__mdt record? Then, the class wouldn't be in the DeveloperName, which has to be unique. Our own Rollup Control record could reference our own Rollup Plugin record that references the RollupLogger class. This way, it wouldn't conflict with the Org Default that ships with the package which we wouldn't use.

But your suggestion sounds more flexible and meets other needs. And there might be reasons why making the class name separate from the DeveloperName isn't ideal (like backward compatibility). We've just scratched the surface of the features of apex-rollup, so I'll follow your lead!

Also, we'll probably switch to the namespaced package anyways. We report our code coverage via gearset which excludes namespaces, and the non-namespaced package causes our coverage % to drop since apex-rollup tests don't install by default (which makes sense to me).

**Reply by jamessimone** [reply]
I'm keen on keeping backward compatibility, but I'll think this one through today and get back to you. Also just a note - I'm sure you saw this but it bears repeating - you can install the [Extra Tests plugin](../tree/main/plugins/ExtraCodeCoverage) to meet the code coverage requirements for Apex Rollup. No worries if you still want to cut over to the namespaced package; just want to ensure you know which options are available to you!

**Reply by pyao-bwc** [reply]
Oh! I did see something about the Extra Tests plugin but didn't look into it carefully enough. I'm glad you mentioned it, that'll help us, thanks!!
